### ISSUE TITLE
RFC: Mutable ASTs vs. Shared Immutable ASTs

### ISSUE THREAD

**Issue by OlivierHamel** [souffle-lang/souffle#2029]
_**State of Affairs**_

The AST/RAM (same issue, just 'AST' hereon) is (mostly[1]) immutable but not shared. Any mutation (i.e. transform) involves a *lot* of cloning / pointer chasing if the node is a non-terminal. Since there's no sharing of immutable nodes, this can turn simple transforms into `O(n^2)` or worse. This is the worst of both worlds and for huge programs, it is painful performance-wise.

**Proposed solutions:**
* mutable ASTs
* shared immutable ASTs
* ??? (<insert alternate proposal here>)

I'm leaning towards mutable ASTs as it probably requires fewer changes than immutable-shared.

_**Mutable ASTs**_
Treat nodes as plain structs; no accessors, can directly modify fields.

Main areas changed: removing accessors & `protected`/`private` modifiers

**Pro:**
* `O(1)` field mutations, including re-parenting
* the node-mapper design already basically treats nodes as mutable
* no accessor boilerplate
* changes consist mainly of removing accessors & `protected`/`private` modifiers

**Con:**
* mutable is always easier to footgun than immutable

_**Shared Immutable AST**_
Go full immutable and change `Own`[2] to something like rust's `Rc<A>` (don't need `shared_ptr`'s atomic ref-counting).

Main areas changed: transforms

**Pro:**
* solves the clone-children problem
* immutable prevents footguns
* can share subtrees (?? dubious utility?)

**Con:**
* distinct subtrees can no longer be distinguished by address
* node-mapping is best done in post-order instead of pre-order (current default)
* node-mapping requires cloning of parent trees to introduce changes (with post-order this is amoratised `O(1)`)

[1] There are spots where a mutable API is exposed (e.g. `Clause`). This is ad-hoc/inconsistent.
[2] It was intended specifically AST node references, not as a general alias for `unique_ptr`.

**Comment by b-scholz** [comment]
This is a good question. We have been thinking about this problem for a while and experimenting with it. 

Note that there is also a difference between RAM and AST. AST permits to change attributes (mainly because of parsing) whereas RAM does not. For RAM, a new node must be created. With attributes I mean fields that are not referencing to children in the AST/RAM trees.

Beside the technical implementation of the IR, I would ask a more high level questions. What are Souffle's IR requirements?

First, I feel one of the most important requirements has been a clear separation of concern between AST, RAM, and execution environment (interpreter/synthesiser). In the past, we did not achieve this separation. 
It took us a long time to clearly separate AST from RAM, and RAM from the execution environments (synthesiser / interpreter).  This separation is absolutely essential for assumptions in transformations (avoiding stale states / convoluted code / etc). I believe that we have achieved this goal mostly. There are still problems:
 -  In `RamNestedOperation` nodes,  we store a profile text for measuring frequencies. That needs be addressed in future so that RAM is truly disconnected from the execution environment. We would need to introduce a new RAM node for expressing frequency profiling behaviour. 
 -  Primitive data-types are not defined for each IR.  In a strict semantic meaning of separation, we would require to have for each IR its own data types.  E.g., `ast_domain`/`ast_number`/`ast_float` etc.; `ram_domain`/`ram_number`/ ....
 - We have been discussing whether we should split the responsibility of AST. Basically, producing an AST for parsing / semantic checks only, and another high-level Datalog IR for transformations/optimisations. The problem is that the AST might be incomplete due to erroneous input programs and may lead to convoluted transformer code as a consequence. In my view, splitting the IR is a large job and it is not clear to me, that splitting the IR is beneficial at the moment. Only with an increased number of transformers/optimisations, it would make sense splitting the AST IR. 
 - The parser performs an AST transformation for disjunctions (cf. ParserDriver) etc.; this needs to be fixed urgently but has huge consequences related to the subsequent analysis/transformation pipeline. 

Second, each IR must be descriptive only - basically, POD (Plain Old Data) and minimal in its nature.  This is not  specific to Souffle - see LLVM, which is a perfect example that  follows this strategy (watch Chris Lattner's talks about LLVM). In the past, we did not have passive / minimal data structures and it hurt us. For example, RAM had state and methods for interpreting programs (SymbolTable, Indexing, etc.) and referred to AST.  However, the latest versions of RAM and AST use passive data structures. Also, each is mostly minimal, i.e., it should only store the minimal amount of information to represent the language construct. For example, a loop statement in RAM should only refer to the loop body and should not describe in addition, which relations are queried in that loop. The responsibility of querying relations in a loop body should be transferred to an analysis. As an outstanding task, I would like to see: 
 - a close inspection of all IR nodes whether they are still PODs and "minimal"
 - a RAM parser that reads RAM programs from a  text file and executes the RAM program  directly. I think that a RAM parser would be useful for unit-testing RAM transformers, and it would be useful for other research projects. 

Third, since the IRs are minimal Souffle requires a mechanism that provides data overlays for its IR (which we call currently analyses) so that additional/derivable information about IR nodes/programs can be computed, cached and queried. The current analysis mechanism relies on re-computation. When the IR changes by a transformer, the information must be re-computed since implementation of incremental analyses updates are most of the time tedious and erroneous. Basically, if a transformer changes the IR, all analyses attached to the program are discarded for a subsequent transformer. This mechanism avoids stale information for transformers/semantic checker. We did not fully achieve this yet:
 -  We had some performance issues in the past, and we create some hidden caches (check utilities in AST/RAM). A more careful analysis is required where we can improve the performance and still have a consistent data overlay mechanism.
 - Can we find a better analysis mechanism that updates/re-computes while executing a transformer? This was the main reason why we had to use hidden caches in the utilities. 

Fourth,  we must have efficient and effective search patterns for analyses and transformers. We developed idioms for these searches. My questions are:
 - Can we provide implementations for current search idioms that perform better and have better readability? I refer to the class of `visit()` functions here?
 - We had problems because the IR nodes did not register all children in the method `getChildNodes()`; some children were forgotten while implementing a node class. Can we use introspection (C++20 and onwards) so that the C++ compiler automatically registers children? Another problem was related to our visitor pattern implementations. Nodes must be registered manually.

Fifth, we require rewriting idioms for the IR. Currently, we use `apply()` and lambdas to rewrite IR trees/sub-trees. The current idioms are hard to write, test, and understand especially for a wider community. Have a look here:
        https://github.com/souffle-lang/souffle/blob/7e1934f145324699eb719aa6d39bace420d40d40/src/ram/transform/ExpandFilter.cpp#L33-L58
    It would be great to have better idioms that support rewriting more verbosely (and have better performance).

Sixth, we require a transformation pipeline that can express complex optimisations.  We have implemented a transformation pipeline with a small DSL (see TranslationUnit/Transformer etc.) permitting sequences and loops. With that we can run optimisations iteratively to build complex optimisation processes. I think this is mostly fine (i.e. functional); perhaps not pretty... We could make the DSL verbatim in form of a text file.


To summarise,  I would prefer discussing the requirements first and double-check where things are really bad in terms of achieving these functional and non-functional requirements (i.e. performance) before talking about concrete implementations details.
